crates/transaction-pool/src/pool/listener.rs

```diff
@@ -35,133 +40,122 @@ impl Stream for TransactionEvents {
+    pub(crate) events: UnboundedReceiver<TransactionEvent>,
+}
+
+impl Stream for AllTransactionsEvents {
+    type Item = TransactionEvent;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.get_mut().events.poll_recv(cx)
+    }
+}
+
 type EventBroadcast = UnboundedSender<TransactionEvent>;
 
 /// A type that broadcasts [`TransactionEvent`] to installed listeners.
 ///
 /// This is essentially a multi-producer, multi-consumer channel where each event is broadcast to
 /// all active receivers.
-#[derive(Debug, Default)]
+#[derive(Default, Debug)]
 pub(crate) struct PoolEventBroadcast {
-    /// All listeners for certain transaction events.
-    broadcasters: HashMap<TxHash, PoolEventBroadcaster>,
+    /// All listeners for all transaction events.
+    all_events_broadcaster: PoolEventBroadcaster,
+    /// All listeners for events for a certain transaction hash.
+    broadcasters_by_hash: HashMap<TxHash, PoolEventBroadcaster>,
 }
 
 impl PoolEventBroadcast {
     /// Calls the broadcast callback with the `PoolEventBroadcaster` that belongs to the hash.
-    fn broadcast_with<F>(&mut self, hash: &TxHash, callback: F)
-    where
-        F: FnOnce(&mut PoolEventBroadcaster),
-    {
-        let is_done = if let Some(sink) = self.broadcasters.get_mut(hash) {
-            callback(sink);
-            sink.is_done()
-        } else {
-            false
-        };
+    fn broadcast_event(&mut self, hash: &TxHash, event: TransactionEventType) {
+        // Broadcast to all listeners for the transaction hash.
+        if let Entry::Occupied(mut sink) = self.broadcasters_by_hash.entry(*hash) {
+            sink.get_mut().broadcast(*hash, event.clone());
 
-        if is_done {
-            self.broadcasters.remove(hash);
+            if sink.get().is_empty() || event.is_final() {
+                sink.remove();
+            }
         }
+
+        // Broadcast to all listeners for all transactions.
+        self.all_events_broadcaster.broadcast(*hash, event);
     }
 
     /// Create a new subscription for the given transaction hash.
     pub(crate) fn subscribe(&mut self, tx_hash: TxHash) -> TransactionEvents {
         let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
 
-        match self.broadcasters.entry(tx_hash) {
+        match self.broadcasters_by_hash.entry(tx_hash) {
             Entry::Occupied(mut entry) => {
                 entry.get_mut().senders.push(tx);
             }
             Entry::Vacant(entry) => {
-                entry.insert(PoolEventBroadcaster { is_done: false, senders: vec![tx] });
+                entry.insert(PoolEventBroadcaster { senders: vec![tx] });
             }
         };
         TransactionEvents { hash: tx_hash, events: rx }
     }
 
+    /// Create a new subscription for all transactions.
+    pub(crate) fn subscribe_all(&mut self) -> AllTransactionsEvents {
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
```

Comment:
I think this is a potential issue because events will be a multiple of all tx in the pool

should make this bounded imo